# Ebay Kleinanzeigen Dataset for Data cleaning Exercise

The dataset was scraped and uploaded to Kaggle. In this exercise, the dataset had been modified as follows:
1. Only 50,000 data points were randomly smapled for the speed.
2. A dirtied version of the dataset to reflect the original status of a scraped dataset.


In [1]:
import numpy as np
import pandas as pd
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [2]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Initial Observation 

After importing to a dataframe: autos, we noticed a few things about this dataset. 
1. There are 20 columns in total, including 5 columns in numeric format and others a mixture of dates, strings and a combination of integers and strings.
2. 15 columns have 50,000 entries while 5 columns have missing entries, null values ranging from (0, 10,000)
3. This dataset takes more than 7.6MB of the memory
4. Some column names contain camelcase instead of using Python's snakecase convention.

Therefore, let's change the column names to better understanding snakecases.

In [3]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [4]:
for c in autos.columns:
    if c == "yearOfRegistration":
        autos.rename({"yearOfRegistration":"registration_year"},inplace=True, axis=1)
    elif c == "monthOfRegistration":
        autos.rename({"monthOfRegistration":"registration_month"},inplace=True, axis=1)
    elif c == "notRepairedDamage":
        autos.rename({"notRepairedDamage":"unrepaired_damage"},inplace=True, axis=1)
    elif c == "dateCreated":
        autos.rename({"dateCreated":"ad_created"},inplace=True, axis=1)
    else:     
        autos.rename({c:c.lower()},inplace=True, axis=1)   

In [5]:
autos.head()

datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicletype  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerps   model   odometer  registration_month fueltype  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrofpictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalcode             lastseen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Clean column names

In the above code block, we have changed the camelcased column names to snakecased while modified four of them to shorter and simpler names.

In [6]:
autos.describe(include="all")

datecrawled         name  seller offertype  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicletype  registration_year  gearbox       powerps  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrofpictures    postalcode             lastseen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In [9]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [10]:
autos["offertype"].value_counts()

Angebot    49999
Gesuch         1
Name: offertype, dtype: int64

In [25]:
autos["nrofpictures"].value_counts()

0    50000
Name: nrofpictures, dtype: int64

### Potential Cleaning

After investigating the descriptive information of the dataset, we have detected a few potential cleaning tasks to be completed.
1. column "seller" and "offertype" both have almost one value. Probably not very meaningful for our analysis. These two columns could be dropped. "nrofpictures" has a mono value of 50000 is not very useful and to be dropped.
2. both "price" and "odometer" columns are essentially numbers but disguised in strings. We should turn them into floats for further analysis.
3. "datecrawled" is by seconds which probably are too detailed to be used for our purpose. Potentially, we could turn this into just year-month or year-month-day or just consider them the same and drop this column
4. Obviously there are some mistakes in the column "registration_year". Most likely any number beyond (1900, 2016) is an error and should be marked as "unknown" or "nan". Meanwhile, column "registration_month" has 0s which also requires attention. 
5. Similarly, column "powerps" being 0 is probably an error and need to be adjusted. 

Let's first transfer the strings to integers for "price" and "odometer".

In [36]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(float)

In [37]:
autos["odometer"]=autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

In [34]:
autos.rename({"odometer":"odometer_km"},inplace=True, axis=1)

In [38]:
autos.head()

datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicletype  registration_year  \
0  privat   Angebot  5000.0  control         bus               2004   
1  privat   Angebot  8500.0  control   limousine               1997   
2  privat   Angebot  8990.0     test   limousine               2009   
3  privat   Angebot  4350.0  control  kleinwagen               2007   
4  privat   Angebot  1350.0     test       kombi               2003   

     gearbox  powerps   model  odometer_km  registration_month fueltype  \
0    manuell      158  andere       150000                   3      lpg   
1  automatik      286     7er       150000                   6   benzin   
2    manuell      102    golf        70000                   7   benzin   
3  automatik       71  fortwo        70000                   6   benzin   
4    manuell        0   focus       150000                   7   benzin   

        brand unrepaired_damage           ad_created  nrofpictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalcode             lastseen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [42]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [47]:
autos["price"].value_counts().sort_index().head(50)


0.0      1421
1.0       156
2.0         3
3.0         1
5.0         2
8.0         1
9.0         1
10.0        7
11.0        2
12.0        3
13.0        2
14.0        1
15.0        2
17.0        3
18.0        1
20.0        4
25.0        5
29.0        1
30.0        7
35.0        1
40.0        6
45.0        4
47.0        1
49.0        4
50.0       49
55.0        2
59.0        1
60.0        9
65.0        5
66.0        1
70.0       10
75.0        5
79.0        1
80.0       15
89.0        1
90.0        5
99.0       19
100.0     134
110.0       3
111.0       2
115.0       2
117.0       1
120.0      39
122.0       1
125.0       8
129.0       1
130.0      15
135.0       1
139.0       1
140.0       9
Name: price, dtype: int64

In [51]:
autos.loc[autos["price"]>10000000, :]

datecrawled                                               name  \
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   

       seller offertype       price   abtest vehicletype  registration_year  \
2897   privat   Angebot  11111111.0     test   limousine               1973   
24384  privat   Angebot  11111111.0     test         NaN               1995   
27371  privat   Angebot  12345678.0  control         NaN               2017   
39377  privat   Angebot  12345678.0  control         NaN               2018   
39705  privat   Angebot  99999999.0  control   limousine               1999   
42221  privat   Angebot  27322222.0  control   limousine               2014   
47598  privat   Angebot  12345678.0  control   limousine               2001   

         gearbox  powerps     model  odometer_km  registration_month fueltype  \
2897     manuell       48    escort        50000                   3   benzin   
24384        NaN        0       NaN       150000                   0      NaN   
27371        NaN       95     punto       150000                   0      NaN   
39377    manuell       95       v40       150000                   6      NaN   
39705  automatik      224  s_klasse       150000                   9   benzin   
42221    manuell      163        c4        40000                   2   diesel   
47598    manuell      101    vectra       150000                   3   benzin   

               brand unrepaired_damage           ad_created  nrofpictures  \
2897            ford              nein  2016-03-12 00:00:00             0   
24384     volkswagen               NaN  2016-03-21 00:00:00             0   
27371           fiat               NaN  2016-03-09 00:00:00             0   
39377          volvo              nein  2016-03-08 00:00:00             0   
39705  mercedes_benz               NaN  2016-03-22 00:00:00             0   
42221        citroen               NaN  2016-03-08 00:00:00             0   
47598           opel              nein  2016-03-31 00:00:00             0   

       postalcode             lastseen  
2897        94469  2016-03-12 22:45:27  
24384       18519  2016-03-21 14:40:18  
27371       96110  2016-03-09 15:45:47  
39377       14542  2016-04-06 23:17:31  
39705       73525  2016-04-06 05:15:30  
42221       76532  2016-03-08 20:39:05  
47598        4356  2016-03-31 18:56:54

In [71]:
autos.drop(autos[autos["price"] > 10000000].index, inplace=True)

In [72]:
autos.loc[autos["price"]>10000000, :]

Empty DataFrame
Columns: [datecrawled, name, seller, offertype, price, abtest, vehicletype, registration_year, gearbox, powerps, model, odometer_km, registration_month, fueltype, brand, unrepaired_damage, ad_created, nrofpictures, postalcode, lastseen]
Index: []

### Dropping rows based on "Price"

Some vehicles have price tags more than 10million and the numbers look artibary, e.g. 12345678. It seems like those are dummy inputs in the system. Therefore, we have decided to delete those 5 records.
